### DISCUSSION TITLE
dns error

### DISCUSSION THREAD

**Discussion by zhen86889** [Kong/kong#12855] [Help] [answered]
2024/01/24 03:32:09 [error] 40#0: *412372185 [lua] job.lua:284: execute(): [timer-ng] failed to run timer unix_timestamp=1706067129907.000000;counter=6761692:meta=debug off: /usr/local/openresty/lualib/resty/dns/resolver.lua:384: bad argument #1 to 'lshift' (number expected, got nil), context: ngx.timer

**Comment by chronolaw** [comment]
@chobits, could you take a look? thanks.

**Reply by chobits** [reply]
It seems that our `timer-ng` has suppressed complete backtrace of runtime error. Might need @zhen86889 add some debug code to get the whole runtime backtrace. So with only this single message, we could not investigate it.

Could you give us a simple step to reproduce.

**Reply by chobits** [reply]
Hi @zhen86889 , or you can modify timer-ng job.lua to enable the complete backtrace and post it here. I wrote a simple patch to help timer-ng to show the complete backtrace, pls test it before using it, ty! 

```
diff --git a/lib/resty/timerng/job.lua b/lib/resty/timerng/job.lua
index 242de14..739291c 100644
--- a/lib/resty/timerng/job.lua
+++ b/lib/resty/timerng/job.lua
@@ -271,8 +271,11 @@ function _M:execute()

     self._running = true

-    local ok, err = pcall(self.callback, ngx_worker_exiting(),
-                          table_unpack(self.argv, 1, self.argc))
+    local ok, err = xpcall(
+      function()
+        self.callback(ngx_worker_exiting(), table_unpack(self.argv, 1, self.argc))
+      end,
+      debug.traceback)

     local finish = stats.finish
```

**Reply by zhen86889** [reply]
According to the provided fix, adjusting job.lua and retrying resulted in the following error:

2024/04/15 06:25:19 [error] 941#0: *23624 [lua] job.lua:286: execute(): [timer-ng] failed to run timer unix_timestamp=1713162319690.000000;counter=531:meta=@/usr/local/share/lua/5.1/kong/resty/dns/client.lua:807:asyncQuery(): /usr/local/openresty/lualib/resty/dns/resolver.lua:384: bad argument #1 to 'lshift' (number expected, got nil)

Steps to reproduce the issue:
1. Configure DNS domain in Kong.
2. For a domain with a length of 60 characters, e.g., xxxxxxxx-xx-xxxxx.xxx-x-xxxxxx.x.xx.xxxx.xxxxxx.xxxxxx.xxxxx, when resolving 45 IPs under the domain, it behaves normally, but when resolving 54 IPs, it encounters an exception.
   For a domain with a length of 59 characters, e.g., xxxxxxxx-xx-xxxx.xxx-x-xxxxxx.x.xx.xxxx.xxxxxx.xxxxxx.xxxxx, when resolving 45 IPs under the domain, it behaves normally, even with 54 IPs.

**Reply by chobits** [reply]
From your steps to reproduce, it seems like lua-resty-dns internal parsing issue. could you try it using this library send query to your domain in your network environment, you could let gpt write a test script for resty. Just want to know if it still exists without kong intergration.

**Reply by zhen86889** [reply]
Hello, according to your suggestion, after testing with lua-resty-dns, I encountered an error. The error message is as follows:

ERROR: /usr/local/openresty/lualib/resty/dns/resolver.lua:384: bad argument #1 to 'lshift' (number expected, got nil)
stack traceback:
        /usr/local/openresty/lualib/resty/dns/resolver.lua:384: in function 'parse_section'
        /usr/local/openresty/lualib/resty/dns/resolver.lua:745: in function 'parse_response'
        /usr/local/openresty/lualib/resty/dns/resolver.lua:909: in function 'query'
        test.lua:19: in function 'file_gen'
        init_worker_by_lua:45: in function <init_worker_by_lua:43>
        [C]: in function 'xpcall'
        init_worker_by_lua:52: in function <init_worker_by_lua:50>

It seems to be an issue with lua-resty-dns. Do you have any suggestions for the next steps?

**Reply by chronolaw** [reply]
Perhaps we should create a fix PR for lua-resty-dns?

**Reply by chobits** [reply]
hi @zhen86889 
Cool!
Could you show your DNS response if there are no sensitive information, like `dig @<dns server ip> <domain>`

And I found some one has filed an issue to lua-resty-dns official: https://github.com/openresty/lua-resty-dns/issues/70

**Reply by zhen86889** [reply]
Due to sensitive information, I've masked part of the content. Here's the translation:

```
nslookup xxxxxxxx-xx-xxxxx.xxx-x-xxxxxx.x.xx.xxxx.xxxxxx.xxxxxx.xxxxx
Server:         xxx.xx.xxx.x
Address:        xxx.xx.xxx.x#53

Name:   xxxxxxxx-xx-xxxxx.xxx-x-xxxxxx.x.xx.xxxx.xxxxxx.xxxxxx.xxxxx
Address: 17.183.214.161
Name:   xxxxxxxx-xx-xxxxx.xxx-x-xxxxxx.x.xx.xxxx.xxxxxx.xxxxxx.xxxxx
Address: 17.183.214.162
Name:   xxxxxxxx-xx-xxxxx.xxx-x-xxxxxx.x.xx.xxxx.xxxxxx.xxxxxx.xxxxx
Address: 17.183.214.163
Name:   xxxxxxxx-xx-xxxxx.xxx-x-xxxxxx.x.xx.xxxx.xxxxxx.xxxxxx.xxxxx
Address: 17.183.214.164
Name:   xxxxxxxx-xx-xxxxx.xxx-x-xxxxxx.x.xx.xxxx.xxxxxx.xxxxxx.xxxxx
Address: 17.183.214.165
Name:   xxxxxxxx-xx-xxxxx.xxx-x-xxxxxx.x.xx.xxxx.xxxxxx.xxxxxx.xxxxx
Address: 17.183.214.166
...
Total of 54 records
```

**Reply by zhen86889** [reply, marked as answer]
Hello, Following the testing method outlined in [https://github.com/openresty/lua-resty-dns/issues/70](url) , we set up a simplified DNS server and were unable to reproduce the issue. Therefore, we shifted our troubleshooting focus to our production DNS service. After analysis, we discovered that the issue was caused by our DNS server truncating the response without setting the `tc` flag when transmitting information that exceeded the UDP packet size limit. Thanks again to the community for your assistance.
